### DISCUSSION TITLE
Authentication using just Backend Functions

### DISCUSSION THREAD

**Discussion by osseonews** [nhost/nhost#2694] [Q&A]
for various reasons, we think we may have to do authentication on the backend of application instead of using the frontend provider with nhost.auth. So we are investigating doing this by just hitting the backend api for Nhost. Have two quick questions:
1. What backend request (docs here https://docs.nhost.io/reference/auth/sign-up-email-and-password), do we use to confirm the user after signup? After signup, they need to confir their email, so they get a  link in their email with a code in the URL. We grab that code and send it to our backend, but how do we check if the code is the same? We don't see any backend functions to actually verify the user. The database stores something called ticket: verifyEmail and hash, but how do we now verify it? Checked the docs to find a verifyEmail backend function, but can't find it.
2. Same question with the magic link. how is the magic link actually confirmed via a backend call, what endpoint is used?

3. Is the benefit of using the frontend, just because all the code is set up already, because it seems like the basic idea here is simple. Login via the Auth service, get back a token, save that token locally (either in cookie or localstorage) and then use that token for requests etc. I see that the frontend has some sort polling set up to constantly generate new tokens of the refresh, but that's also something that we dn't really need.

**Comment by osseonews** [comment]
I was looking at the emails that get sent out t and the verification happens via a link to nhost.run server which then redirects back to our site? So there is no way to verify the email on our own server? I'm also a bit confused by this auth flow. Even if we use the nhostprovider, how is the client actually informed that there was a verification, if this is done on a different server. Usually with the auth libraries we work with, you grab the code from the URL and see if matches the code in the database. But this is not happening here, so how does this even work. nhost.run can't set a cookie on our server with the token.

**Comment by dbarrosop** [comment]
> the verification happens via a link to nhost.run server which then redirects back to our site?

correct

> how is the client actually informed that there was a verification

the redirect will either include the `refreshToken` or an error code. You can try yourself by going to that url with a valid code and a fake one.

> Usually with the auth libraries we work with, you grab the code from the URL and see if matches the code in the database

That's what the server does when you visit the URL. Well, that plus a couple of other checks but the most important one is validating the code is valid.

> nhost.run can't set a cookie on our server with the token

Correct, that will happen on the frontend application so if you are implementing this workflow yourself you will have to do this yourself, otherwise, our SDK handles this automatically.

> Same question with the magic link. how is the magic link actually confirmed via a backend call, what endpoint is used?

Magic link works the same way.

> Is the benefit of using the frontend, just because all the code is set up already, because it seems like the basic idea here is simple...

If you mean our SDK, that's correct. Our SDK implements all the common workflows but you are free to implement them yourself. Everything is available through the auth's API and there is no need to use our SDK if you don't want to or if you are using a language for which we don't haven an SDK (i.e. our CLI, written in go, will just make calls to the auth service directly)

If you have more questions don't hesitate to let us know.

**Reply by osseonews** [reply]
Thanks I get all that now, but what is the function or API endpoint to actually verify the email from the URL? I can't find it anywhere. If we wanted to verify the email on our own server instead of redirecting them to nhost.run, how can we do that? Searching Nhost Github repo I found this:  - 744fd69: remove `nhost.auth.verifyEmail`  Theres's a /verify endpoint in hasura-auth, but the sdk is not even using it as..." Where is the nhost.auth.verifyEmail now?

**Reply by dbarrosop** [reply]
The endpoint to verify emails is the URL you get in the email. If you want to verify the emails from your backend/function you will need to receive the ticket in your backend/function somehow (i.e. by tweaking the email templates) and then calling the url yourself with the ticket.

**Reply by dbarrosop** [reply]
Not exactly the same but this is a similar use case (even though it uses our SDK):

https://nhost.io/blog/protect-magic-links-from-email-clients

In your case instead of the frontend app being in the middle to protect the links it'd be your backend/functions.

**Reply by osseonews** [reply]
Thanks, that link is very useful. Actually, the problem you are bringing up is exactly why I think you need to have OTP in email, as I brought up on this issue: https://github.com/nhost/nhost/issues/2651. I really don't see any of the major auth libs using Magic links anymore. It's all 6-digit code sent to email or a phone for verifications.  Links in emails are just sort of antiquated and many internal systems will strip them for security reasons. You end up with a complex system of code with workarounds, as mentioned in your post. It works, of course, but there is no need for it. A 6 digit code is way easier. Really this is the only thing missing in your auth.

**Reply by osseonews** [reply]
BTW, quick question, in your post: https://nhost.io/blog/protect-magic-links-from-email-clients, what are the variables you can use for email verification if you want to use this method? It seems like for email verification there is an extra parameter required, when I inspected the link sent in the email. I checked out the template for email verification and all that it had was: {{link}} - what are the parameters for the email template though that are required to construct the link so that when we redirect from our React app, we are sending the correct link to nhost?

**Reply by dbarrosop** [reply]
You can find the template variables here:

https://docs.nhost.io/guides/auth/email-templates#template-variables
